### PR TITLE
Fixed #33245 -- Made django.utils.html.urlize() thread-safe.

### PR DESCRIPTION
**Pull request by tim-mccurrach**
`urlize()` is not currently thread-safe. This fixes that.

See [ticket](https://code.djangoproject.com/ticket/33245)

### CODE REVIEW

File: django/utils/html.py [hunk 1]
```diff
@@ -250,25 +250,28 @@ def __call__(self, text, trim_url_limit=None, nofollow=False, autoescape=False):
...
+                word,
+                trim_url_limit=trim_url_limit,
+                nofollow=nofollow,
+                autoescape=autoescape,
+                safe_input=safe_input
+            ) for word in words
         ])
 
-    def handle_word(self, word):
+    def handle_word(self, word, trim_url_limit, nofollow, autoescape, safe_input):
```
Comment (claudep) [inline django/utils/html.py:266]: I think I would define the additional parameters as keyword parameter, as in `__call__ `.

BTW, thanks for fixing this thread issue I created!

Comment (tim-mccurrach) [reply django/utils/html.py]: Thanks, yes that seems sensible 👍

**Comment by felixxm** [comment]
@tim-mccurrach Thanks :+1:

File: django/utils/html.py [hunk 2]
```diff
@@ -294,20 +299,20 @@ def handle_word(self, word):
...
+        elif safe_input:
             return mark_safe(word)
-        elif self.autoescape:
+        elif autoescape:
             return escape(word)
         return word
 
-    def trim_url(self, x):
-        if self.trim_url_limit is None or len(x) <= self.trim_url_limit:
+    def trim_url(self, x, trim_url_limit):
```
Comment (ngnpope) [inline django/utils/html.py:312]: The `trim_url_` prefix for the new argument seems unnecessary here.
Also, if we're expecting it to be passed as a keyword, let's enforce that.

```suggestion
    def trim_url(self, x, *, limit=None):
```

File: django/utils/html.py [hunk 3]
```diff
@@ -283,8 +288,8 @@ def handle_word(self, word):
                 nofollow_attr = ''
             # Make link.
             if url:
-                trimmed = self.trim_url(middle)
-                if self.autoescape and not self.safe_input:
+                trimmed = self.trim_url(middle, trim_url_limit=trim_url_limit)
```
Comment (ngnpope) [inline django/utils/html.py:291]:
```suggestion
                trimmed = self.trim_url(middle, limit=trim_url_limit)
```

File: django/utils/html.py [hunk 4]
```diff
@@ -250,25 +250,30 @@ def __call__(self, text, trim_url_limit=None, nofollow=False, autoescape=False):
...
+                trim_url_limit=trim_url_limit,
+                nofollow=nofollow,
+                autoescape=autoescape,
+            ) for word in words
         ])
 
-    def handle_word(self, word):
+    def handle_word(
+        self, word, safe_input, trim_url_limit=None, nofollow=False, autoescape=False,
+    ):
```
Comment (ngnpope) [inline django/utils/html.py:268]: Ensure that the new arguments are keyword only:

```suggestion
    def handle_word(
        self, word, *, safe_input, trim_url_limit=None, nofollow=False, autoescape=False,
    ):
```

**Review by ngnpope — COMMENTED** [review]
Sorry to be late to the review party.

I only had some minor nitpicks, but feel free to ignore if you're against them...

**Comment by tim-mccurrach** [comment]
> Sorry to be late to the review party.
> 
> I only had some minor nitpicks, but feel free to ignore if you're against them...

Thanks, @ngnpope. I'm certainly not against them. What's the protocol for further changes on something that's already merged. Just open a new PR that references the same ticket?

**Comment by felixxm** [comment]
> Just open a new PR that references the same ticket?

Yes.

**Pull request merged**